GOSSIP.md

```diff
@@ -34,45 +35,252 @@ also used to determine which node should keep its initiated connection open in
+
+Regardless of the peer already being in a pool or not, some information
+about it is updated when it is received through the gossip protocol:
+
+- Last gossiped time:
+
+  This is used to downgrade peers that have not been gossiped for a long time.
+
+- (optional) The counter of the number of time it has been seen:
+
+  Every time the peer is received through the gossip protocol, a time-based
+  windowed counter is incremented. e.g. a counter per day for the last N days.
+
+  This is used to estimate the duplication of the peer in the unverified pool
+  to favor the frequently seen peers during selection. The duplication is
+  bounded and decreases exponentially.
+
+#### Unverified Pool
+
+The unverified pool is composed of 1024 buckets of up to 64 peers, resulting in
+a maximum capacity of 65536 peers.
+
+To prevent a rogue node to fill it with compromised peers, it uses the address
+of the node gossiping the peers to select a subset of the bucket; then the added
+peer address is used to select the bucket it belongs to. A secret known only by
+the node is used to randomize the selection process so it cannot be predicted
+by the attacker.
+
+If a peer is not already in the verified pool, the steps to add it to the
+unverified pool are:
+
+1. If already in the unverified pool, all references are first removed to
+prevent uncontrolled duplications.
+2. A subset of 64 buckets are selected based on the secret and the IP of the
+node that gossiped the peer; this limits the part of the pool that can be
+changed by any given rogue node IP.
+3. From this subset, 16 buckets are selected based on the secret and the IP of
+the peer to be added; this randomizes the peer distribution preventing the
+prediction of the way peers will be evicted.
+4. From these 16 buckets, 8 are selected based on the secret and the port of the
+added peer; this reduces the collisions between peers that share the same IP.
+5. (optional) In function of the frequency the peer has seen (See
+[General Peer Status](#general-peer-status)) it is added to 1 up to 8 of these
+buckets; this favor peers that were gossiped more frequently when picking a new
+one to connect to and possible upgrade to the verified pool.
+6. If the bucket the peer has to be added to is full, one existing peer has to
+be evicted. This is done by first cleaning all the peers that weren't gossiped
+for a certain amount of time, then if the bucket is still full, by selecting a
+random peer with a bias favoring peers that were added the longest time ago.
+7. The eventually evicted peer is completely removed from the pool.
+
+Only the IP of the source node is used to select a subset of the pool because
+IP is an expensive resource while ports are comparatively cheap.
+
+See [Bucket Selection](#bucket-selection) for more details on how the buckets
+are selected from the secret and other discriminators.
+
+#### Verified Pool
+
+The verified pool is composed of 256 buckets of up to 64 peers, resulting in a
```

Comment:
This is far too many! As a comparison there are a total of ~16000 ethereum nodes at the moment...

Reply:
I understand, maybe 128 buckets are enough. Keep in mind that due to the way bucket are selected buckets will be filled in a linear way so the maximum should be bigger than the target.